### ISSUE TITLE
IdentifiedRef.load broke in 3.0.0-rc.1

### ISSUE THREAD

**Issue by stephenh** [mikro-orm/mikro-orm#267]
When I bump from `3.0.0-rc` to `3.0.0-rc.1`, this code is no longer working:

```typescript
    async email(root) {
      const client = await root.load();
      return client.email;
    },
```

Where root is an `IdentifiedReference`.

I don't see a tag for `rc` vs. `rc.1` to do a compare and scan for likely changes, so I don't really have much more information than "it started happening in 3.0.0-rc.1"

(I had been on `3.0.0-rc.3` and it had the broken behavior as well; it looks like `3.0.0-rc` it was working fine and then regressed in `3.0.0-rc.1`.

**Comment by B4nan** [comment]
What does it mean, not working? It does not load the entity, so `client.email` is undefined? What is output of `root.isInitialized()` after the `load` call? Tests for this are passing so the general functionality should be ok.

**Comment by B4nan** [comment]
This should be the list of changes between `rc.0` and `rc.1`:

https://github.com/mikro-orm/mikro-orm/compare/v3.0.0-rc...b7b7a46

**Comment by stephenh** [comment]
Hey, yeah, sorry, that was vague. `client.email` is coming back undefined enough those its set in the db. For some reason isInitialized is still false after the load call... I'll work on a better reproduction. Maybe I'm doing something dumb.

**Comment by B4nan** [comment]
One possible problem might be with request context - how do you set that up? Do you handle it via `RequestContext` helper, or manually via `em.fork()`? 
One of those changes was to correct the internal `entity.__em` instance when merging the entity. This EM instance is then used in `entity.init()` call, which is exactly what happens when you call `Reference.load()` if the entity is not yet intialized. 

You could try to manually revert this change in the compiled code of ORM in your project, here is the commit: https://github.com/mikro-orm/mikro-orm/commit/72f4525e6d2048691fb0b9875a025ae5dfea7c5f

**Comment by stephenh** [comment]
Huh, interesting. You're right, if I undo that commit just locally in my node_modules/ directory, it works again...

I'm manually calling `em.fork()`; the Apollo server has a `context` method that it calls at the beginning of each request:

```typescript
export async function createServer(context: AppContext): Promise<ApolloServer> {
  return new ApolloServer({
    typeDefs,
    resolvers,
    extensions: [() => errorLogging()],
    context: async ({ req }): Promise<Context> => {
      const token = req ? req.headers.authorization : undefined;
      return {
        ...context,
        user: await context.authenticator(token),
        em: context.orm.em.fork(),
        req,
      };
    },
  });
}
```

Where `context` is like an application-wide context object that has a shared `orm: MikroORM` instance that each request forks from.

**Comment by B4nan** [comment]
Sounds like the entity you have in the parameter (`root`) is loaded with wrong EM (not the request fork).

Maybe it has to do something with your dataprovider implementation, or you are not using that here?

**Comment by stephenh** [comment]
Hm, I've done some logging and you are right that `IdentifiedRef.load()` is merging into the wrong EM, but I'm pretty sure I'm using the right one.

It looks like when the ref is created, a shell of the Entity is created here:

https://github.com/mikro-orm/mikro-orm/blob/master/lib/entity/EntityFactory.ts#L80

Which has a pointed to the non-forked/root EM  `entity.__em`, see the `useContext: true` in this screenshot:

![image](https://user-images.githubusercontent.com/6401/70678857-edc48300-1c58-11ea-9fd7-15ecdeaabd99.png)

And then a few attributes down on the em in the screenshot is a dummy `__uniqueid` tag I added that identifies it as the very first EM created.

So, I kind of get that is the default, but AFAICT the entity/ref is never corrected to have the right EM.

I.e. see the screenshot and the console.logs:

![image](https://user-images.githubusercontent.com/6401/70679026-8bb84d80-1c59-11ea-92e4-4fc7b7d0415f.png)

I'm using "EM 3" to do the `find(Project)` call but the refs come back pointing at "EM 1".

My guess is that the commit you found did not cause the "EM 1" to be wrong, and if anything AFAICT when that line runs, the EM is likely fixed, however it's too late, the ref's entity (which was wrongly hooked up to EM 1) doesn't see the data that is populated into the "EM 3"'s unit of work.

Here's my project model:

```typescript
@Entity({ collection: "projects" })
export class Project implements IdEntity<Project> {
  @PrimaryKey()
  id: number;

  @ManyToOne({ entity: () => Client, wrappedReference: true })
  client: IdentifiedReference<Client>;
}
```

**Comment by stephenh** [comment]
See https://github.com/stephenh/mikro-orm/commit/7a7656f749c06f8c242c4380e0c966982d06e2c5 for a reproduction. I had to change the `publisher` over to an `IdentifiedReference` and update the related code changes, so it's probably not a reproduction you'd want to merge directly as a canonical test case for it (I'll defer to you for the best place to cover this in the test suite), but should at least show what is happening.

**Comment by B4nan** [comment]
Thanks for the repro, much appreciated! I will try to look into this soon, should have some spare time during weekend if not sooner. 

Regarding your finding, yes, the root EM should be there by default, and if one merges the entity to different one, it should be overridden. This is mainly because when you create new instance via constructor, I do not know to which EM the entity should belong, so during merge this gets fixed. Maybe the issue is that we replace only the EM instance on parent entity, but not on its children/relations. Or as you said, we need to do it sooner. Or maybe actually combination of those two things. 

Btw the `useContext` flag is there just for `RequestContext` helper, as you are not using it, that should have no effect on anything.

**Comment by stephenh** [comment]
> yes, the root EM should be there by default

I'm worried that will create more confusion/bugs than it is worth; i.e. it treats the root EM as special, which I don't think it should be, b/c for basically any server-side environment, you specifically don't want the root EM to accidentally sneak into places, like it has done here, and get used by accident.

I.e. while you're poking around, I'd personally suggest trying to clip that "the root EM is set by default on entities" all together, and then fix all of the places that break b/c of that, and I assert that in the process of doing that, you'd probably end up fixing the "oh right, refs aren't setting the em when they should" bug that I seem to be tripping over (and likely several others we don't know about yet).

> Btw the useContext flag is there just for RequestContext helper, as you are not using it,
> that should have no effect on anything.

Sure, I was just mentioning that since it was the first "ah ha" I had insofar as "...wait a minute, that's not the EM that I want, b/c mine should have `useContext: flag`", b/c otherwise the EMs look very opaque in the debugger and it's hard to tell whether you have the right one or not.

**Comment by B4nan** [comment]
Agreed, it just adds to the wtf factor. 

Will need to double check, but I think the only need for this is to support entity.assign(), which automatically creates references from PKs. But we can add optional em parameter and throw when not set already on the entity.

**Comment by B4nan** [comment]
So the problem was that the correction was happening only for the merged entity, but it was not cascaded to its relations. 

Tricky part was to actually remove the need for always present EM instance, but it cracked that as well. Will fix the rest of the bugs first before releasing next RC, probably during weekend.

**Comment by stephenh** [comment]
Great, thanks!

**Comment by B4nan** [comment]
Fixed in `3.0.0-rc.4`.
